flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.changelog.fs;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.runtime.state.StreamStateHandle;
+import org.apache.flink.runtime.state.changelog.SequenceNumber;
+import org.apache.flink.runtime.state.changelog.StateChange;
+import org.apache.flink.runtime.state.changelog.StateChangelogHandleStreamImpl;
+import org.apache.flink.runtime.state.changelog.StateChangelogWriter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.changelog.fs.StateChangeSet.Status.PENDING;
+import static org.apache.flink.runtime.concurrent.FutureUtils.combineAll;
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+@NotThreadSafe
```

Comment:
I don't think that every implementation has to be not thread safe.

Reply:
But it's not property of the implementation, but rather of the interface, as the interface defines the treading model.

Reply:
Could you explain why the "threading model defined by this interface" is not thread-safe?

Reply:
`StateChangelogWriter` is an SPI interface. This SPI defines how this interface will be used, including the threading model. If the base interface is used in a multi threading setting and the implementations are expected to be `@ThreadSafe`, implementations can not change this contract. If the base interface is used in a single threaded model (or always under some kind of global lock) stating `@NotThreadSafe` on the implementation is a no-op.

I mean interface and SPI defines the contract how is it being used, so stating `@NotThreadSafe` or `@ThreadSafe` on the concrete classes is either a no-op or a sign of something potentially wrong happening.

Reply:
Thanks for the clarification.
I don't think that interface/SPI **must** define this. For example, `java.util.Map` **doesn't** - because implementations can be thread-safe and non-thread-safe. `BlockingQueue` **does** - because it defines operations specifically for the multi-threaded invironment.